src/StreamingParser.php

```diff
@@ -0,0 +1,362 @@
+<?php
+namespace DemandwareXml;
+
+use Generator;
+use SimpleXMLElement;
+use XMLReader;
+
+class StreamingParser
+{
+    protected $file;
+    protected $skipAttributes = false;
+
+    public function __construct(string $file)
+    {
+        $this->file = $file;
+    }
+
+    public function skipAttributes(bool $skipAttributes)
+    {
+        $this->skipAttributes = $skipAttributes;
+    }
+
+    // Validation without a schema is allowed for photos as they don't have one.
+    public function validate($useSchema = true): bool
+    {
+        if (! file_exists($this->file)) {
+            throw new XmlException('XML file does not exist: ' . basename($this->file));
+        } elseif (! is_readable($this->file)) {
+            throw new XmlException('XML file is not readable: ' . basename($this->file));
+        }
+
+        $previousErrors = libxml_use_internal_errors(true);
+        libxml_clear_errors();
+
+        set_error_handler(function (int $severity, string $message, string $file, int $line) {
+            throw new XmlException($message . ' in ' . basename($file) . ' on line ' . $line, $severity);
+        });
+
+        try {
+            $reader = new XMLReader;
+            $reader->open($this->file);
+
+            if ($useSchema) {
+                $reader->setSchema(realpath(__DIR__ . '/../xsd/catalog.xsd'));
+            }
+
+            while ($reader->read());
+
+            $reader->close();
+
+            $errors = libxml_get_errors();
+
+            if (count($errors) > 0) {
+                throw $this->libXmlErrorToException(reset($errors));
+            }
+        } finally {
+            libxml_use_internal_errors($previousErrors);
+            libxml_clear_errors();
+            restore_error_handler();
+        }
+
+        return true;
+    }
+
+    protected function libXmlErrorToException($error): XMLException
+    {
+        switch ($error->level) {
+            case LIBXML_ERR_WARNING:
+                $level = 'Warning';
+                break;
+
+            case LIBXML_ERR_ERROR:
+                $level = 'Error';
+                break;
+
+            case LIBXML_ERR_FATAL:
+                $level = 'Fatal';
+                break;
+        }
+
+        return new XmlException($level . ': ' . trim($error->message) . ' in ' . basename($error->file) . ' on line ' . $error->line . ' column ' . $error->column, $error->code);
+    }
+
+    protected function parseNodes(string $node): Generator
```

Comment:
Will this always return a `Generator`, or could it return `void` (if there are no matching nodes, for example)?